### PR TITLE
Update visual prompting pipeline for multi-label zero-shot learning support

### PR DESCRIPTION
**Pull request by sungchul1**
### Summary

This PR includes below updates for multi-label support:
- [x] Add `ZeroShotPromptType`
- [x] Migrate polygons or bitmap masks into prompts
- [x] Integrate `inference_step`

** The table below shows reproduced results based on dice score.
- There is a performance difference between before and after this update because:
    - polygon and bitmap mask that were ground truth are used as prompts as well (bbox/point -> bbox/point/polygon/bitmap mask)
    - predicted masks are not multiplied with scores anymore
    - after multi-label enablement, benchmark performance can be changed one more because the current version has only considered single-label inference

| model | task | before (52fef3d7) | after (a811a114) |
| -- | -- | -- | -- |
| sam_tiny_vit (fine-tuning) | test | 0.91 | 0.92 |
|  | export | 1.00 | 0.92 |
|  | optimize | 0.99 | 0.92 |
| sam_vit_b (fine-tuning) | test | 0.93 | 0.94 |
|  | export | 1.00 | 0.94 |
|  | optimize | 1.00 | 0.93 |
| sam_tiny_vit (zero-shot) | test | 0.59 | 0.53 |
|  | export | 0.59 | 0.58 |
|  | optimize | 0.55 | 0.52 |
| sam_vit_b (zero-shot) | test | 0.32 | 0.29 |
|  | export | 0.33 | 0.30 |
|  | optimize | 0.29 | 0.26 |

### Checklist

- [x] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [x] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Comment by sovrasov** [comment]
@sungchul2 could you provide ZSL results on reduced coco (torch+OV) to prove there's no regression?

File: src/otx/core/data/entity/visual_prompting.py [hunk 1]
```diff
@@ -6,19 +6,22 @@
...
 from otx.core.types.task import OTXTaskType
 
 if TYPE_CHECKING:
-    from datumaro import Polygon as dmPolygon
     from torch import LongTensor
-    from torchvision.tv_tensors import BoundingBoxes as tvBoundingBoxes
-    from torchvision.tv_tensors import Mask as tvMask
+
+
+ZeroShotPromptType = Type[tvBoundingBoxes | Points | tvMask | dmPolygon]
```
Comment (eunwoosh) [inline src/otx/core/data/entity/visual_prompting.py:24]:
```suggestion
ZeroShotPromptType = tvBoundingBoxes | Points | tvMask | dmPolygon
```
Isn't it unnecessary to use `Type` in this case?

File: src/otx/core/data/entity/visual_prompting.py [hunk 2]
```diff
@@ -127,26 +130,16 @@ class VisualPromptingBatchPredEntity(OTXBatchPredEntity, VisualPromptingBatchDat
     """Data entity to represent model output predictions for visual prompting task."""
 
 
-@dataclass
-class ZeroShotVisualPromptingLabel:
```
Comment (eunwoosh) [inline src/otx/core/data/entity/visual_prompting.py:131]: It's just a question. why is it unnecessary now?

File: src/otx/core/model/visual_prompting.py [hunk 3]
```diff
@@ -596,7 +584,7 @@ def _set_label_info(self, _: LabelInfoTypes) -> None:
     def get_dummy_input(self, batch_size: int = 1) -> ZeroShotVisualPromptingBatchDataEntity:
         """Returns a dummy input for ZSL VPT model."""
         images = [torch.rand(3, *self.input_size) for _ in range(batch_size)]
-        labels = [ZeroShotVisualPromptingLabel(prompts=torch.LongTensor([0]))] * batch_size
+        labels = [{"points": torch.LongTensor([0] * batch_size)}] * batch_size
```
Comment (eunwoosh) [inline src/otx/core/model/visual_prompting.py:587]:
```suggestion
        labels = [{"points": torch.LongTensor([0] * batch_size)}] * batch_size
```
I'm not sure but is it ok to multiply batch_size twice?

File: src/otx/algo/visual_prompting/sam.py [hunk 4]
```diff
@@ -56,30 +56,37 @@ class CommonSettingMixin:
         "vit_l": "https://dl.fbaipublicfiles.com/segment_anything/sam_vit_l_0b3195.pth",
         "vit_h": "https://dl.fbaipublicfiles.com/segment_anything/sam_vit_h_4b8939.pth",
     }
-    load_state_dict: Callable[[dict[str, Tensor]], None]
 
-    def load_checkpoint(self, load_from: str | None) -> None:
+    def load_state_dict(
+        self,
+        state_dict: dict[str, Any] | None = None,
```
Comment (eunwoosh) [inline src/otx/algo/visual_prompting/sam.py:62]: Could you update docstring for arugment?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work :) I left some comments. Please take a look.

File: src/otx/core/data/entity/visual_prompting.py (hunk 2 above)
Comment (sungchul1) [reply src/otx/core/data/entity/visual_prompting.py:131]: Before, I tried to manage prompts as different key-value pairs and through this update prompts are combined into a list of prompts. As a result, this data class is unnecessary anymore.

File: src/otx/core/model/visual_prompting.py (hunk 3 above)
Comment (sungchul1) [reply src/otx/core/model/visual_prompting.py:587]: It doesn't matter to use it twice because it is just to create dummy inputs for exportation.

File: src/otx/core/data/entity/visual_prompting.py (hunk 1 above)
Comment (sungchul1) [reply src/otx/core/data/entity/visual_prompting.py:24]: Mypy was upset without this.
```
error: Variable "otx.core.data.entity.visual_prompting.ZeroShotPromptType" is not valid as a type  [valid-type]
error: ZeroShotPromptType? has no attribute "pin_memory"  [attr-defined]
error: ZeroShotPromptType? has no attribute "cpu"  [attr-defined]
```

File: src/otx/algo/visual_prompting/sam.py (hunk 4 above)
Comment (sungchul1) [reply src/otx/algo/visual_prompting/sam.py:62]: Done at 2280c01

**Review by kprokofi — CHANGES_REQUESTED** [review]
postpone to merge PR before 2.2/2.3 is finalized

**Review by kprokofi — APPROVED** [review]

**Review by eunwoosh — APPROVED** [review]
